File: Cardano ObjC/Cardano ObjC/ViewController.m
//
//  ViewController.m
//  Cardano ObjC
//
//  Created by Glen Yi on 2014-08-21.
//  Copyright (c) 2014 On The Pursuit. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController
            
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)goClick:(UIButton *)sender {
    NSInteger limit = self.limitTextField.text.integerValue;
    if (limit>=3) {
        NSDate *timestamp = [NSDate date];
        
        self.timeLabel.text = @"Calculating...";
        self.limitTextField.enabled = self.goButton.enabled = NO;
        
        [self calculateTripletsWithLimit:limit completion:^(NSInteger count) {
            self.tripletsLabel.text = [NSString stringWithFormat:@"%li", count];
            NSInteger seconds = -[timestamp timeIntervalSinceNow];
            self.timeLabel.text = [NSString stringWithFormat:@"%li", seconds];
            
            self.limitTextField.enabled = self.goButton.enabled = YES;
            NSLog(@"Found %li triplets in %li seconds!", (long)count, (long)seconds);
        }];
    }
}

- (BOOL)isCardanoTriplet:(NSInteger)a b:(NSInteger)b c:(NSInteger)c {
    double bc = b * sqrt(c);
    double cardano = cbrt(a + bc) + cbrt(a - bc);
    
    // Float errors... if there's a better way then please let me know
    return (cardano < 1.00000000000001 && cardano > 0.99999999999999);
}

- (void)calculateTripletsWithLimit:(NSInteger)limit completion:(void(^)(NSInteger count))completion {
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        NSInteger count = 0;
        
        for (int a=1; a<limit-2; a++) {
            for (int b=1; b<limit-a-1; b++) {
                for (int c=1; c<limit-a-b; c++) {
                    if ([self isCardanoTriplet:a b:b c:c]) {
                        count++;
                        NSLog(@"%li: (%i, %i, %i)", (long)count, a, b, c);
                    }
                }
            }
        }

        dispatch_sync(dispatch_get_main_queue(), ^{
            completion(count);
        });
    });
}

@end

File: Cardano Swift/Cardano Swift/ViewController.swift
//
//  ViewController.swift
//  Cardano Swift
//
//  Created by Glen Yi on 2014-08-21.
//  Copyright (c) 2014 On The Pursuit. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
                            
    @IBOutlet var limitTextField: UITextField!
    @IBOutlet var goButton: UIButton!
    @IBOutlet var tripletsLabel: UILabel!
    @IBOutlet var timeLabel: UILabel!
    
    typealias Triplet = (a: Int, b: Int, c: Int)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func goClick(sender: UIButton) {
        // Find all cardano triplets
        if let limit = limitTextField.text.toInt() {
            if limit >= 3 {
                // Retrieve current time
                let timestamp = NSDate()
                
                // UI updates
                self.timeLabel.text = "Calculating..."
                self.goButton.enabled = false
                self.limitTextField.enabled = false
                
                self.calculateTriplets(limit, { count in
                    // Update triplets and time in seconds
                    self.tripletsLabel.text = "\(count)"
                    let seconds = -timestamp.timeIntervalSinceNow
                    self.timeLabel.text = "\(Int(seconds))"
                    
                    // UI updates
                    self.goButton.enabled = true
                    self.limitTextField.enabled = true
                    println("Found \(count) triplets in \(Int(seconds)) seconds!")
                } )
            }
        }
    }
    
    func isCardanoTriplet(triplet: Triplet) -> Bool {
        let bc = Double(triplet.b) * sqrt(Double(triplet.c))
        let cardano = cbrt(Double(triplet.a)+bc) + cbrt(Double(triplet.a)-bc)
        
        // Float errors... if there's a better way then please let me know
        let isCardano = (cardano < 1.00000000000001 && cardano > 0.99999999999999)
        
        return isCardano
    }
    
    func calculateTriplets(limit: Int, completion: (Int)->()) {
        // Calculate the triplets asynchronously
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), {
            var count = 0
            
            // Here is where most of the computing time happens, a brute force solution.
            // I'm sure there are some very elegant mathematical simplifications here but that's beyond the scope of this exercise.
            for a in 1...limit-2 {
                for b in 1...limit-a-1 {
                    for c in 1...(limit-a-b) {
                        let triplet = (a, b, c)
                        if self.isCardanoTriplet(triplet) {
                            count++
                            println("\(count): \(triplet)")
                        }
                    }
                }
            }
            
            // Return results on main thread
            dispatch_sync(dispatch_get_main_queue(), {
                completion(count)
            })
        })
    }

}

